File: driver/python/pysequoiadb/collectionspace.py
#   Copyright (C) 2012-2014 SequoiaDB Ltd.
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

"""Module of collectionspace for python driver of SequoiaDB
"""

try:
   import sdbcs
except ImportError:
   try:
      import libsdbcs
   except ImportError:
      raise Exception("cannot find C module file: sdbcs")

import bson
import pysequoiadb
from pysequoiadb import ( static_object,
                          default_host,
                          default_svcname,
                          default_user,
                          default_psw )
from pysequoiadb.collection import collection
from pysequoiadb.cursor import cursor
from pysequoiadb import error
from pysequoiadb.common import const
from pysequoiadb.error import SequoiaDBError

class collectionspace(object):
   """CollectionSpace for SequoiaDB
   
   All operation need deal with the error code returned first, if it has. 
   Every error code is not SDB_OK(or 0), it means something error has appeared,
   and user should deal with it according the meaning of error code printed.

   @version: execute to get version
             >>> import pysequoiadb
             >>> print pysequoiadb.get_version()

   @notice : The dict of built-in Python is hashed and non-ordered. so the
             element in dict may not the order we make it. we make a dict and
             print it like this:
             ...
             >>> a = {"avg_age":24, "major":"computer science"}
             >>> a
             >>> {'major': 'computer science', 'avg_age': 24}
             ...
             the elements order it is not we make it!!
             therefore, we use bson.SON to make the order-sensitive dict if the
             order is important such as operations in "$sort", "$group",
             "split_by_condition", "aggregate","create_collection"...
             In every scene which the order is important, please make it using
             bson.SON and list. It is a subclass of built-in dict
             and order-sensitive
   """
   
   def __init__(self):
      """invoked when a new object is producted.

      """
      #'cs' is short for collection space
      try:
         self._cs = sdbcs.create_cs()
      except SystemError:
         pysequoiadb.check_error(const.SDM_OOM)
         raise SequoiaDBError

   def __del__(self):
      """delete a object existed.

      """
      if self._cs is not None:
         rc = sdbcs.release_cs(self._cs)
         pysequoiadb.check_error(rc)
         self._cs = None

   def __repr__(self):

      return "Collection Space %s" %(self.get_collection_space_name())

   def __getattr__(self, name):
      """support client.cs to access to collection.

         eg.
         cc = client()
         cs = cc.test
         cl = cs.test_cl  # access to collection named 'test_cl'

         and we should pass '__members__' and '__methods__',
         becasue dir(cc) will invoke __getattr__("__members__") and
         __getattr__("__methods__").

         if success, a collection object will be returned, or None.
      """
      if '__members__' == name or '__methods__' == name:
         pass
      else:
         cl = collection()
         rc = sdbcs.get_collection(self._cs, name, cl._cl)
         pysequoiadb.check_error(rc)
         if const.SDB_OK != rc:
            del cl;
            cl = None
         return cl

   def __getitem__(self, name):
      """support [] to access to collection.
      
         eg.
         cc = client()
         cs = cc['test']
         cl = cs['test_cl']# access to collection named 'test_cl'.
      """
      return self.__getattr__(name)

   def get_collection(self, cl_name):
      """Get the named collection.
         
      Parameters:
         Name         Type     Info:
         cl_name      str      The full name of the collection..
      Return values:
         Success: SDB_OK  and  a collection object of query
         Fail   : Others  and  None
      """
      if not isinstance(cl_name, basestring):
         raise TypeError("collection must be an instance of basestring")

      cl = collection()
      rc = sdbcs.get_collection(self._cs, cl_name, cl._cl)
      pysequoiadb.check_error(rc)

      if const.SDB_OK != rc:
         del cl
         cl = None

      return rc, cl

   def create_collection(self, cl_name, options = static_object):
      """create a collection using name and options.

      Parameters:
         Name      Type     Info:
         cl_name   str      The collection name.
         options   str      The options for creating collection, including
                                  "ShardingKey", "ReplSize", "IsMainCL" and
                                  "Compressed" informations, no options, if None.
      Return values:
         Success: SDB_OK  and  a collection object created
         Fail   : Others  and  None

      """
      if not isinstance(cl_name, basestring):
         raise TypeError("collection must be an instance of basestring")

      bson_options = None
      if options is not None:
         bson_options = bson.BSON.encode(options)

      cl = collection()
      if options is None:
         rc = sdbcs.create_collection(self._cs, cl_name, cl._cl)
      else:
         rc = sdbcs.create_collection_use_opt(self._cs, cl_name,
                                              bson_options, cl._cl)
      pysequoiadb.check_error(rc)

      if const.SDB_OK != rc:
         del cl
         cl = None

      return rc, cl

   def drop_collection(self, cl_name):
      """Drop the specified collection in current collection space.

      Parameters:
         Name      Type     Info:
         cl_name   str      The collection name.
      Return values:
         Success: SDB_OK
         Fail   : Others
      """
      if not isinstance(cl_name, basestring):
         raise TypeError("collection must be an instance of basestring")

      rc = sdbcs.drop_collection(self._cs, cl_name)
      pysequoiadb.check_error(rc)

      return rc

   def get_collection_space_name(self):
      """Get the current collection space name.

      Parameters:
         Name      Type     Info:
         N/A
      Return values:
         The name of current collection space.
      """
      _, cs_name = sdbcs.get_collection_space_name(self._cs)
      pysequoiadb.check_error(_)

      return cs_name

File: driver/java/src/main/java/com/sequoiadb/base/CollectionSpace.java
/*
 * Copyright 2017 SequoiaDB Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

package com.sequoiadb.base;

import com.sequoiadb.exception.BaseException;
import com.sequoiadb.exception.SDBError;
import com.sequoiadb.message.request.AdminRequest;
import com.sequoiadb.message.response.SdbReply;
import org.bson.BSONObject;
import org.bson.BasicBSONObject;

import java.util.ArrayList;
import java.util.List;

/**
 * @class CollectionSpace
 * @brief Database operation interfaces of collection space.
 */
public class CollectionSpace {
    private String name;
    private Sequoiadb sequoiadb;

    /**
     * @return The collection space name
     * @fn String getName()
     * @brief Return the name of current collection space.
     */
    public String getName() {
        return name;
    }

    /**
     * @return Sequoiadb object
     * @fn Sequoiadb getSequoiadb()
     * @brief Return the Sequoiadb instance of current collection space belong to.
     */
    public Sequoiadb getSequoiadb() {
        return sequoiadb;
    }

    /**
     * @param sequoiadb Sequoiadb handle
     * @param name      Collection space name
     * @fn CollectionSpace(Sequoiadb sequoiadb, String name)
     * @brief Constructor
     */
    CollectionSpace(Sequoiadb sequoiadb, String name) {
        this.name = name;
        this.sequoiadb = sequoiadb;
    }

    /**
     * @param collectionName The collection name
     * @return The collection object or null for collection not exist
     * @throws com.sequoiadb.exception.BaseException
     * @fn DBCollection getCollection(String collectionName)
     * @brief Get the named collection
     */
    public DBCollection getCollection(String collectionName)
        throws BaseException {
        // get cl from cache
        String collectionFullName = name + "." + collectionName;
        if (sequoiadb.fetchCache(collectionFullName)) {
            return new DBCollection(sequoiadb, this, collectionName);
        }

        // no need to upsert/remove cache here,
        // for "isCollectionExist" has do that
        if (isCollectionExist(collectionName)) {
            return new DBCollection(sequoiadb, this, collectionName);
        } else {
            throw new BaseException(SDBError.SDB_DMS_NOTEXIST, collectionName);
        }
    }

    /**
     * @param colName The collection name
     * @return True if collection existed or False if not existed
     * @throws com.sequoiadb.exception.BaseException
     * @fn boolean isCollectionExist(String colName)
     * @brief Verify the existence of collection in current collection space
     */
    public boolean isCollectionExist(String colName) throws BaseException {
        String collectionFullName = name + "." + colName;

        BSONObject obj = new BasicBSONObject();
        obj.put(SdbConstants.FIELD_NAME_NAME, collectionFullName);

        AdminRequest request = new AdminRequest(AdminCommand.TEST_CL, obj);
        SdbReply response = sequoiadb.requestAndResponse(request);

        int flag = response.getFlag();
        if (flag == 0) {
            sequoiadb.upsertCache(collectionFullName);
            return true;
        } else if (flag == SDBError.SDB_DMS_NOTEXIST.getErrorCode()) {
            sequoiadb.removeCache(collectionFullName);
            return false;
        } else {
            sequoiadb.throwIfError(response);
            return false; // make compiler happy
        }
    }

    /**
     * @return A list of collection names
     * @throws com.sequoiadb.exception.BaseException
     * @fn List<String> getCollectionNames()
     * @brief Get all the collection names of current collection space
     */
    public List<String> getCollectionNames() throws BaseException {
        List<String> collectionNames = new ArrayList<String>();
        List<String> colNames = sequoiadb.getCollectionNames();
        if ((colNames != null) && (colNames.size() != 0)) {
            for (String col : colNames) {
                if (col.startsWith(name + ".")) {
                    collectionNames.add(col);
                }
            }
        }
        return collectionNames;
    }

    /**
     * @param collectionName The collection name
     * @throws com.sequoiadb.exception.BaseException
     * @fn void createCollection(String collectionName)
     * @brief Create the named collection in current collection space
     */
    public DBCollection createCollection(String collectionName)
        throws BaseException {
        return createCollection(collectionName, null);
    }

    /**
     * @param collectionName The collection name
     * @param options        The options for creating collection, including
     *                       "ShardingKey", "ReplSize", "IsMainCL" and "Compressed" informations,
     *                       no options, if null
     * @return the created DBCollection
     * @throws com.sequoiadb.exception.BaseException
     * @fn DBCollection createCollection(String collectionName, BSONObject
     * options)
     * @brief Create collection by options
     */
    public DBCollection createCollection(String collectionName,
                                         BSONObject options) {
        if (isCollectionExist(collectionName)) {
            throw new BaseException(SDBError.SDB_DMS_EXIST, collectionName);
        }

        String collectionFullName = name + "." + collectionName;

        BSONObject obj = new BasicBSONObject();
        obj.put(SdbConstants.FIELD_NAME_NAME, collectionFullName);
        if (options != null) {
            obj.putAll(options);
        }

        AdminRequest request = new AdminRequest(AdminCommand.CREATE_CL, obj);
        SdbReply response = sequoiadb.requestAndResponse(request);
        String msg = "collection = " + collectionFullName + ", options = " + options;
        sequoiadb.throwIfError(response, msg);
        sequoiadb.upsertCache(collectionFullName);
        return new DBCollection(sequoiadb, this, collectionName);
    }

    /**
     * @return void
     * @throws com.sequoiadb.exception.BaseException
     * @fn void drop()
     * @brief Drop current collectionSpace
     * @see com.sequoiadb.base.Sequoiadb.dropCollectionSpace
     * @deprecated the method will be deprecated in version 2.x, use Sequoiadb.dropCollectionSpace instead
     */
    public void drop() throws BaseException {
        sequoiadb.dropCollectionSpace(this.name);
    }

    /**
     * @param collectionName The collection name
     * @throws com.sequoiadb.exception.BaseException
     * @fn void dropCollection(String collectionName)
     * @brief Remove the named collection of current collection space
     */
    public void dropCollection(String collectionName) throws BaseException {
        if (!isCollectionExist(collectionName)) {
            throw new BaseException(SDBError.SDB_DMS_NOTEXIST, collectionName);
        }

        String collectionFullName = name + "." + collectionName;

        BSONObject obj = new BasicBSONObject();
        obj.put(SdbConstants.FIELD_NAME_NAME, collectionFullName);

        AdminRequest request = new AdminRequest(AdminCommand.DROP_CL, obj);
        SdbReply response = sequoiadb.requestAndResponse(request);
        sequoiadb.throwIfError(response, collectionName);
        sequoiadb.removeCache(collectionFullName);
    }
}
